wakuv2/filter_manager.go

```diff
@@ -53,225 +28,95 @@ type FilterEvent struct {
-	} else {
-		sub = subDetails[0]
-		logger.Debug("filter subscription success", zap.Stringer("peer", sub.PeerID), zap.String("pubsubTopic", contentFilter.PubsubTopic), zap.Strings("contentTopics", contentFilter.ContentTopicsList()))
-	}
-
-	success := err == nil
-	mgr.eventChan <- FilterEvent{eventType: FilterEventSubscribeResult, filterID: filterID, tempID: tempID, sub: sub, success: success}
-}
-
-func (mgr *FilterManager) unsubscribeFromFilter(filterID string, sub *subscription.SubscriptionDetails) {
-	mgr.logger.Debug("filter unsubscribe from filter node", zap.String("filterId", filterID), zap.String("subId", sub.ID), zap.Stringer("peer", sub.PeerID))
-	// Unsubscribe on light node
-	ctx, cancel := context.WithTimeout(mgr.ctx, requestTimeout)
-	defer cancel()
-	_, err := mgr.node.FilterLightnode().UnsubscribeWithSubscription(ctx, sub)
-
-	if err != nil {
-		mgr.logger.Warn("could not unsubscribe wakuv2 filter for peer", zap.String("filterId", filterID), zap.String("subId", sub.ID), zap.Error(err))
-	}
+	go mgr.filterLifecycle(ctx, filterID, contentFilter)
 
-	success := err == nil
-	mgr.eventChan <- FilterEvent{eventType: FilterEventUnsubscribeResult, filterID: filterID, success: success, sub: sub}
 }
-
-// Check whether each of the installed filters
-// has enough alive subscriptions to peers
-func (mgr *FilterManager) pingPeers() {
-	mgr.logger.Debug("filter pingPeers")
-
-	distinctPeers := make(map[peer.ID]struct{})
-	for filterID, subs := range mgr.filterSubs {
-		logger := mgr.logger.With(zap.String("filterId", filterID))
-		nilSubsCnt := 0
-		for _, s := range subs {
-			if s == nil {
-				nilSubsCnt++
-			}
-		}
-		logger.Debug("filter ping peers", zap.Int("len", len(subs)), zap.Int("len(nilSubs)", nilSubsCnt))
-		if len(subs) < mgr.settings.MinPeersForFilter {
-			// Trigger full resubscribe
-			logger.Debug("filter ping peers not enough subs")
-			go func(filterID string) {
-				mgr.eventChan <- FilterEvent{eventType: FilterEventPingResult, filterID: filterID, success: false}
-			}(filterID)
-		}
-		for _, sub := range subs {
-			if sub == nil {
-				// Skip temp subs
-				continue
-			}
-			_, found := distinctPeers[sub.PeerID]
-			if found {
-				continue
-			}
-			distinctPeers[sub.PeerID] = struct{}{}
-			logger.Debug("filter ping peer", zap.Stringer("peerId", sub.PeerID))
-			go func(sub *subscription.SubscriptionDetails) {
-				err := mgr.isFilterSubAlive(sub)
-				alive := err == nil
-
-				if alive {
-					logger.Debug("filter aliveness check succeeded", zap.Stringer("peerId", sub.PeerID))
-				} else {
-					logger.Debug("filter aliveness check failed", zap.Stringer("peerId", sub.PeerID), zap.Error(err))
-				}
-				mgr.eventChan <- FilterEvent{eventType: FilterEventPingResult, peerID: sub.PeerID, success: alive}
-			}(sub)
-		}
+func (mgr *FilterManager) removeFilter(filterID string) {
+	cancel, ok := mgr.filters[filterID]
+	if ok {
+		delete(mgr.filters, filterID)
+		// close goroutine running filterLifecycle
+		cancel()
```

Comment:
Shouldn't unsubscribe be called for subscription associated withthis particular filter?

Reply:
It will be called via `cancel()`, as filterLifecycle listens to `ctx.Done()` channel. Once it emits an event, `unsubscribe` is called (line 96).